azurelinuxagent/common/protocol/goal_state.py

```diff
@@ -58,42 +60,46 @@ def __init__(self, wire_client, full_goal_state=False, base_incarnation=None):
-        role_instance = find(xml_doc, "RoleInstance")
-        self.role_instance_id = findtext(role_instance, "InstanceId")
-        role_config = find(role_instance, "Configuration")
-        self.role_config_name = findtext(role_config, "ConfigName")
-        container = find(xml_doc, "Container")
-        self.container_id = findtext(container, "ContainerId")
-        lbprobe_ports = find(xml_doc, "LBProbePorts")
-        self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
-
-        AgentGlobals.update_container_id(self.container_id)
-
-        fetch_full_goal_state = False
-        if full_goal_state:
-            fetch_full_goal_state = True
-            reason = 'force update'
-        elif base_incarnation is not None and self.incarnation != base_incarnation:
-            fetch_full_goal_state = True
-            reason = 'new incarnation'
-
-        if not fetch_full_goal_state:
-            self.hosting_env = None
-            self.shared_conf = None
-            self.certs = None
-            self.ext_conf = None
-            self.remote_access = None
-            return
-
-        logger.info('Fetching new goal state [incarnation {0} ({1})]', self.incarnation, reason)
+        try:
+            uri = GOAL_STATE_URI.format(wire_client.get_endpoint())
+            self.xml_text = wire_client.fetch_config(uri, wire_client.get_header())
+            xml_doc = parse_doc(self.xml_text)
+
+            self.incarnation = findtext(xml_doc, "Incarnation")
+            self.expected_state = findtext(xml_doc, "ExpectedState")
+            role_instance = find(xml_doc, "RoleInstance")
+            self.role_instance_id = findtext(role_instance, "InstanceId")
+            role_config = find(role_instance, "Configuration")
+            self.role_config_name = findtext(role_config, "ConfigName")
+            container = find(xml_doc, "Container")
+            self.container_id = findtext(container, "ContainerId")
+            lbprobe_ports = find(xml_doc, "LBProbePorts")
+            self.load_balancer_probe_port = findtext(lbprobe_ports, "Port")
+
+            AgentGlobals.update_container_id(self.container_id)
+
+            fetch_full_goal_state = False
+            if full_goal_state:
+                fetch_full_goal_state = True
+                reason = 'force update'
+            elif base_incarnation is not None and self.incarnation != base_incarnation:
+                fetch_full_goal_state = True
+                reason = 'new incarnation'
+
+            if not fetch_full_goal_state:
+                self.hosting_env = None
+                self.shared_conf = None
+                self.certs = None
+                self.ext_conf = None
+                self.remote_access = None
+                return
+        except Exception as exception:
+            # We don't log the error here since fetching the goal state is done every few seconds
+            raise ProtocolError(msg="Error fetching goal state", inner=exception)
```

Comment:
Would it be useful to add a traceback here? Might be useful in figuring out what exactly failed when trying to fetch GS

Reply:
I thought you were adding the traceback to error messages? (one way to do that could be in the conversion to string of the base class for exceptions)

Reply:
btw - since this is just raising the exception, the traceback shouln't be added here, but in whoever handles the exception

Reply:
and another note:

    one way to do that could be in the conversion to string of the base class for exceptions)

may lead to bad (but useful) formatting. An alternative would be to create a function in the logger to log an exception. That function could also add traceback for the inner exceptions.

Reply:
>I thought you were adding the traceback to error messages? 

I was primarily talking about doing it on a case to case basis, but I like your idea better of adding it for every error.

> An alternative would be to create a function in the logger to log an exception.

That would be useful for sure,  maybe reuse the logger.error? But anyways not sure when I'll get time to work on that. For now do you wanna add the trace here until we have a better solution in place?

Reply:
The trace should not be added here, that should be done at the place where the exception is handled.